### DISCUSSION TITLE
How to run float16 on CUDAExecutionProvider

### DISCUSSION THREAD

**Discussion by FrancescoSaverioZuppichini** [microsoft/onnxruntime#13145] [General]
Like title.

Thanks!

Cheers

**Comment by tianleiwu** [comment]
CUDA EP supports float16, you need make sure your model uses float16 (like weights shall be in float16) first. 

You can use https://github.com/microsoft/onnxconverter-common/blob/master/onnxconverter_common/float16.py to convert model from float to float16.

**Comment by FrancescoSaverioZuppichini** [comment]
Hi @tianleiwu , thanks a lot! Do you think converting my models weights to fp16 in pytorch before exporting is sufficient?

**Comment by tianleiwu** [comment]
@FrancescoSaverioZuppichini, I recommend export fp32 models, then run graph optimization and finally convert to fp16. It is because some optimization requires that (extra Cast nodes might block some graph optimizations).

**Reply by FrancescoSaverioZuppichini** [reply]
Thanks @tianleiwu , any snippet/tutorial I can use?

**Reply by tianleiwu** [reply]
Example snippet for huggingface BERT model: https://github.com/microsoft/onnxruntime/blob/335b62bde6cfc5fa88cb199ad8f42aeb400857c0/onnxruntime/python/tools/transformers/onnx_exporter.py#L244-L263

**Reply by FrancescoSaverioZuppichini** [reply]
Thanks! I'll check it out
